### PR TITLE
Multiple target node training with per node type decoder

### PR DESCRIPTION
**Pull request by CongWeilin**
*Issue #, if available:*

*Description of changes:*

A decoder is created for each node type. This PR also include functions that create per node type loss function, but haven't been used in `create_builtin_node_gnn_model` yet.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gsf.py [hunk 1]
```diff
@@ -131,14 +131,16 @@ def create_builtin_node_gnn_model(g, config, train_task):
...
         Whether this model is used for training.
+    share_decoder: bool
+        Whether this model share decoder for different node types.
 
     Returns
     -------
     GSgnnModel : The GNN model.
     """
-    return create_builtin_node_model(g, config, train_task)
+    return create_builtin_node_model(g, config, train_task, share_decoder)
```
Comment (CongWeilin) [inline python/graphstorm/gsf.py:141]: If we can add `share_decoder` as part of config, then user can directly create this model by setting model arguments.

File: python/graphstorm/model/gnn.py [hunk 2]
```diff
@@ -395,6 +412,22 @@ def set_loss_func(self, loss_fn):
                 'The loss function should be the class of nn.Module.'
         self._loss_fn = loss_fn
 
+    def set_loss_func_dict(self, loss_fns):
```
Comment (CongWeilin) [inline python/graphstorm/model/gnn.py:415]: I just add this function here, but I did this function when creating graphstorm's model.

File: python/graphstorm/gsf.py [hunk 3]
```diff
@@ -161,18 +165,36 @@ def create_builtin_node_model(g, config, train_task):
...
+                                                else model.node_input_encoder.out_dims,
+                                               config.num_classes,
+                                               config.multilabel))
+        else:
+            decoder = {}
+            for ntype in config.target_ntype:
+                decoder[ntype] = EntityClassifier(model.gnn_encoder.out_dims \
+                                                if model.gnn_encoder is not None \
+                                                else model.node_input_encoder.out_dims,
+                                               config.num_classes,
```
Comment (zheng-da) [inline python/graphstorm/gsf.py:180]: do we need to extend this? different tasks may have different numbers of classes.

File: python/graphstorm/gsf.py [hunk 4]
```diff
@@ -161,18 +165,36 @@ def create_builtin_node_model(g, config, train_task):
...
+                                               config.multilabel)
```
Comment (zheng-da) [inline python/graphstorm/gsf.py:181]: similarly to this one.

File: python/graphstorm/model/gnn.py [hunk 5]
```diff
@@ -383,6 +383,23 @@ def set_decoder(self, decoder):
                     + 'match the input dimension of the decoder.'
         self._decoder = decoder
 
+    def set_decoder_dict(self, decoders):
```
Comment (zheng-da) [inline python/graphstorm/model/gnn.py:386]: instead of adding a new method, can we reuse `set_decoder` and have it to support a dict of decoders?

File: python/graphstorm/model/gnn.py (hunk 2 above)
Comment (zheng-da) [reply python/graphstorm/model/gnn.py:415]: the same comment as above

Comment (zheng-da) [reply python/graphstorm/model/gnn.py:415]: this function isn't used to set different loss functions for different node types.

File: python/graphstorm/model/gnn.py (hunk 5 above)
Comment (CongWeilin) [reply python/graphstorm/model/gnn.py]: merged in new commit.

File: python/graphstorm/gsf.py (hunk 1 above)
Comment (CongWeilin) [reply python/graphstorm/gsf.py]: solved

File: python/graphstorm/model/gnn.py (hunk 2 above)
Comment (CongWeilin) [reply python/graphstorm/model/gnn.py]: solved. In the update, variables `config.num_classess`, `config.multilabel`, `config.multilabel_weights`, and `config.imbalance_class_weights` can be defined as a per node type dictionary.

However, the current implementation only support node classification but does not support node regression, because those variables are not used in regression task in graphstorm.

File: python/graphstorm/config/argument.py [hunk 6]
```diff
@@ -1195,21 +1207,40 @@ def multilabel_weights(self):
...
-                assert w >= 0., "multilabel weights can not be negative values"
-            assert len(weights) == self.num_classes, \
-                "Each class must have an assigned weight"
-
-            return th.tensor(weights)
+        if hasattr(self, "_num_classes") and isinstance(self.num_classes, dict):
+            if hasattr(self, "_multilabel_weights"):
+                ntype_weights = {}
+                for ntype in self.num_classes:
+                    assert self.multilabel[ntype] is True, \
```
Comment (zheng-da) [inline python/graphstorm/config/argument.py:1214]: should we save the `multilabel` result somewhere first? otherwise, it will trigger computations in multilabel every time.

File: python/graphstorm/config/argument.py [hunk 7]
```diff
@@ -1195,21 +1207,40 @@ def multilabel_weights(self):
...
+                        "Must be a multi-label classification task."
+                    try:
+                        weights = self._multilabel_weights[ntype].split(",")
+                        weights = [float(w) for w in weights]
+                    except Exception: # pylint: disable=broad-except
+                        assert False, "The weights should in following format 0.1,0.2,0.3,0.1,0.0"
```
Comment (zheng-da) [inline python/graphstorm/config/argument.py:1220]: can we raise an error instead of using assert? it looks weird.

File: python/graphstorm/config/argument.py [hunk 8]
```diff
@@ -1195,21 +1207,40 @@ def multilabel_weights(self):
...
+                assert self.multilabel is True, "Must be a multi-label classification task."
+                try:
+                    weights = self._multilabel_weights.split(",")
+                    weights = [float(w) for w in weights]
+                except Exception: # pylint: disable=broad-except
+                    assert False, "The weights should in following format 0.1,0.2,0.3,0.1,0.0"
+                for w in weights:
+                    assert w >= 0., "multilabel weights can not be negative values"
+                assert len(weights) == self.num_classes, \
+                    "Each class must have an assigned weight"
```
Comment (zheng-da) [inline python/graphstorm/config/argument.py:1239]: can we wrap the checks here in a function? so we don't need to duplicate the code above.

Comment (zheng-da) [reply python/graphstorm/config/argument.py:1239]: the same comments apply to other places as well.

File: python/graphstorm/run/gsgnn_np/gsgnn_np.py [hunk 9]
```diff
@@ -35,13 +35,22 @@ def get_evaluator(config):
...
+                                    config.early_stop_rounds,
+                                    config.early_stop_strategy)
+        else:
+            return GSgnnAccEvaluator(config.eval_frequency,
+                                    config.eval_metric,
+                                    config.multilabel,
+                                    config.use_early_stop,
+                                    config.early_stop_burnin_rounds,
+                                    config.early_stop_rounds,
+                                    config.early_stop_strategy)
```
Comment (zheng-da) [inline python/graphstorm/run/gsgnn_np/gsgnn_np.py:53]: the only difference is `multilabel`? if so, can we write it like this:
```
multilabel = config.multilabel[config.eval_target_ntype] if isinstance(config.multilabel, dict) else config.multilabel
```

Comment (CongWeilin) [reply python/graphstorm/run/gsgnn_np/gsgnn_np.py]: solved

File: python/graphstorm/config/argument.py (hunk 8 above)
Comment (CongWeilin) [reply python/graphstorm/config/argument.py]: solved

File: python/graphstorm/config/argument.py (hunk 7 above)
Comment (CongWeilin) [reply python/graphstorm/config/argument.py]: solved

File: python/graphstorm/config/argument.py (hunk 6 above)
Comment (CongWeilin) [reply python/graphstorm/config/argument.py]: solved

**Comment by CongWeilin** [comment]
integrate test is provided:

- (step 1) A multi-target node version of movieLens dataset is created in `test/data_gen/create_data.sh` and  `data_gen/process_movielens.py`

- (step 2) A multi-target node single-label classification test case is created in `test/graphstorm-nc/test.sh`

- (step 3) A multi-target node multilabel classification movielens is generated using `test/data_gen/gen_multilabel.py`. Then,  a multi-target node multilabel classification test case is created in `test/graphstorm-nc/test.sh`

File: docs/source/advanced/advanced-usages.rst [hunk 10]
```diff
@@ -16,8 +16,51 @@ GraphStorm provides supports to avoid theses problems:
...
+
+- Evaluate on single node type: During evuation, the users can set a single node type for evaluation. For example, by setting ``eval_target_ntype:  movie``, we will only perform evaluation on "movie" node type. Our current implementation only support evaluating on a single node type.
+
+- Per target node type decoder: The users may also want to use a different decoder on each node type, where the output dimension for each decoder maybe different. We can achieve this by setting ``num_classes`` in model config YAML file. For example, by setting ``num_classes`` as following, GraphStorm will create a decoder with output dimension as 3 for movie node type, and a decoder with output dimension as 7 for user node type.
+
+  .. code-block:: yaml
+
+    num_classes:
+      movie:  3
+      user:  7
```
Comment (zheng-da) [inline docs/source/advanced/advanced-usages.rst:48]: can we still use `num_classes: 7` to specify the number of classes for one target node type?

Comment (zheng-da) [reply docs/source/advanced/advanced-usages.rst:48]: the same question goes to other arguments as well.

Comment (zheng-da) [reply docs/source/advanced/advanced-usages.rst:48]: if not, we are changing our interface significantly.

File: python/graphstorm/config/argument.py [hunk 11]
```diff
@@ -1302,21 +1335,37 @@ def imbalance_class_weights(self):
...
-            assert len(weights) == self.num_classes, \
+            assert len(weights) == num_classes, \
                 "Each class must have an assigned weight"
-
             return th.tensor(weights)
 
-        return None
+        if hasattr(self, "_num_classes") and isinstance(self.num_classes, dict):
+            if hasattr(self, "_imbalance_class_weights"):
+                assert isinstance(self._imbalance_class_weights, dict)
```
Comment (zheng-da) [inline python/graphstorm/config/argument.py:1353]: please provide an error message.

File: docs/source/advanced/advanced-usages.rst (hunk 10 above)
Comment (CongWeilin) [reply docs/source/advanced/advanced-usages.rst:48]: > can we still use `num_classes: 7` to specify the number of classes for one target node type?

Yes, we can still use this command for single node type, same for other arguments. The above instruction is speficially for more than 1 node types.

Comment (CongWeilin) [reply docs/source/advanced/advanced-usages.rst:48]: Actually, most of the integrate test cases and unit test cases are single node type, and we don't have to change their commands.

File: python/graphstorm/config/argument.py (hunk 11 above)
Comment (CongWeilin) [reply python/graphstorm/config/argument.py]: solved.

**Review by zheng-da — APPROVED** [review]
LGTM.

**Comment by zheng-da** [comment]
@classicsong can you take a look at this PR?

File: python/graphstorm/config/argument.py [hunk 12]
```diff
@@ -1302,21 +1335,38 @@ def imbalance_class_weights(self):
...
+        if hasattr(self, "_num_classes") and isinstance(self.num_classes, dict):
+            if hasattr(self, "_imbalance_class_weights"):
+                assert isinstance(self._imbalance_class_weights, dict), \
+                    print('The imbalance_class_weights should be dictionary')
+                num_classes = self.num_classes
+                imbalance_class_weights = self._imbalance_class_weights
+                ntype_weights = {}
+                for ntype in num_classes:
+                    ntype_weights[ntype] = check_imbalance_class_weights(
+                        imbalance_class_weights[ntype],
```
Comment (classicsong) [inline python/graphstorm/config/argument.py:1360]: What if some node types have imbalance class weights and some node types do not?

File: python/graphstorm/config/argument.py [hunk 13]
```diff
@@ -1341,6 +1391,24 @@ def target_ntype(self):
...
+        # pylint: disable=no-member
+        if hasattr(self, "_eval_target_ntype"):
+            assert isinstance(self._eval_target_ntype, str), \
+                "Now we only support single ntype evaluation"
+            return self._eval_target_ntype
+        else:
+            if isinstance(self.target_ntype, str):
+                return self.target_ntype
+            elif isinstance(self.target_ntype, list):
+                # (wlcong) Now only support single ntype evaluation
```
Comment (classicsong) [inline python/graphstorm/config/argument.py:1407]: Can you add a warning here?

Comment (CongWeilin) [reply python/graphstorm/config/argument.py:1413]: solved.

File: python/graphstorm/config/argument.py (hunk 12 above)
Comment (CongWeilin) [reply python/graphstorm/config/argument.py]: solved.
